File: 2570. Merge Two 2D Arrays by Summing Values.js
// 2570. Merge Two 2D Arrays by Summing Values
// LEARN => NUM GROUPING NUM MERGING MERGE TO_H GROUP_BY COMBINE ARRAY ARRAY MERGING ARRAY TO HASH 
nums1 = [[1,2],[2,3],[4,5]]
nums2 = [[1,4],[3,2],[4,1]]
/**
 * @param {number[][]} nums1
 * @param {number[][]} nums2
 * @return {number[][]}
 */
//MY SOL
var mergeArrays = function(nums1, nums2) {
  const mergedMap = new Map();
  
  [...nums1, ...nums2].forEach(([key, value]) => {
    if (!mergedMap.has(key)) {
      mergedMap.set(key, []);
    }
    mergedMap.get(key).push(value);
  });
  
  const transformedValues = new Map([...mergedMap.entries()].map(([key, value]) => [key, value.reduce((acc, curr) => acc + curr, 0)]));
  const sortedEntries = [...transformedValues.entries()].sort((a, b) => a[0] - b[0]);
  return sortedEntries;
}
console.log(mergeArrays(nums1,nums2))

//SOL 2 | CONCISE
var mergeArrays = function(nums1, nums2) {
  const mergedMap = new Map();
  [...nums1, ...nums2].forEach(([k, v]) => {
      mergedMap.set(k, (mergedMap.get(k) || 0) + v);
  });
  return [...mergedMap.entries()].sort((a, b) => a[0] - b[0]);
}
console.log(mergeArrays(nums1,nums2))

File: 2570. Merge Two 2D Arrays by Summing Values.py
#* 2570. Merge Two 2D Arrays by Summing Values
#* LEARN => NUM GROUPING NUM MERGING MERGE TO_H GROUP_BY COMBINE ARRAY ARRAY MERGING ARRAY TO HASH 
nums1 = [[1,2],[2,3],[4,5]]
nums2 = [[1,4],[3,2],[4,1]]
from typing import List
from collections import defaultdict
#* SOL
class Solution:
    def mergeArrays(self, nums1: List[List[int]], nums2: List[List[int]]) -> List[List[int]]:
        merged_dict = defaultdict(int)
        for key, value in nums1 + nums2:
            merged_dict[key] += value

        result = sorted(merged_dict.items())
        return result
    
print(Solution().mergeArrays(nums1,nums2))
        
#*
class Solution:
    def mergeArrays(self, nums1: List[List[int]], nums2: List[List[int]]) -> List[List[int]]:
      merged_dict = {}
      for k, v in nums1:
          merged_dict[k] = merged_dict.get(k, 0) + v
      for k, v in nums2:
          merged_dict[k] = merged_dict.get(k, 0) + v
      return sorted(merged_dict.items())
    
print(Solution().mergeArrays(nums1,nums2))
    

